### PR TITLE
[CSSimplify] Fix sub-pattern matching logic related closure and enum element

### PR DESCRIPTION
**Pull request by takeshi-1000**
This PR fix constraint sytem logic around closures where sub-pattern matching as part of the enum element matching. 
When you compile the code below in swift 5.7, error occur, which message is `type of expression is ambiguous without more context`, though error don't occur in swift 5.5 and swift 5.6.1. So this seems degression in swift 5.7.
```
enum TestType {
    case foo
    case bar(Bool, (a: String, String))
}

func test(type: TestType) -> String {
    let str: String = {
        switch type {
        case .foo:
            return ""
        case .bar(_, (let a, _)):
            return a
        }
    }()
    
    return str
}
```

I could no find any direct information on this subject. It's a similar issue . https://github.com/apple/swift/issues/60806

### CODE REVIEW

**Comment by takeshi-1000** [comment]
@swift-ci please smoke test

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp:2074]: This check should only happen if `path.size() > 2`

**Comment by xedin** [comment]
Thank you for the fix, @takeshi-1000! I left a couple of comments inline.

File: lib/Sema/CSSimplify.cpp
Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp]: Thank you for confirmation.
The original code wasn't good enough for N-level pattern matching, so I'm trying to fix it.

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2074]: Why `5`? Should this logic be more dynamic e.g. find the index for `LocatorPathElt::PatternMatch` and then look through function arguments and tuple elements from that index forward instead of using specific indexes like `3` or `4`? To me even seems a more stable approach for `N` levels of nesting and we could even merge the logic above in there as well.
Note I'm not a reviewer so this is more an opinion that maybe is worth considering =]

Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp:2074]: Thanks, @LucianoPAlmeida :) . I totally agree with you. I will try to fix it.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:2074]: I wonder if simply checking that there is a `PatternMatch` element is in the path is good enough here to determine that location belongs to a pattern matching context.

Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp:2074]: @xedin 
Would it make more sense to define a static value like the current fix?
Or is it that the current fix looks weird?

I tried to change it to use dynamic variables a while ago, but the current modification seemed to be fine considering the case where the pattern matching element was not included.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:2074]: These constants are usually a sign that something is not right. I am wondering whether iterating over a path to find first `PatternMatch` element would actually be sufficient here instead of all the checks we currently have which seem all be too narrow.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:2074]: > I am wondering whether iterating over a path to find first `PatternMatch` element would actually be sufficient here instead of all the checks we currently have which seem all be too narrow.

That makes sense. I was also noting that we are dropping `TupleType` information from locator path but not `TupleElement`. Either way, if we only find first `PatternMatch` we can remove that logic entirely. Even if the narrowing down to `LocatorPathElt::FunctionArgument` is needed we can check the next index after finding the last `PatternMatch` position in the path which also makes this dropping not needed.

Comment (xedin) [reply lib/Sema/CSSimplify.cpp:2074]: Sounds like a plan!

Comment (takeshi-p0601) [reply lib/Sema/CSSimplify.cpp]: Sorry for the late reply.
Thank you everyone for checking and following me.
I tried to fix it, so please check it.

Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp]: Sorry for the late reply.
Thank you everyone for checking and following me.
I tried to fix it, so please check it.

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2248]: Left over from debugging?

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2071]: We could simplify that to 
```suggestion
              auto it = llvm::find_if(path, [](LocatorPathElt &elt) {
        return elt.is<LocatorPathElt::PatternMatch>();
      });
      if (it == path.end())
        return false; 
```
Also, was checking `LocatorPathElt::FunctionArgument` and `LocatorPathElt::TupleElement` really needed? Or  only returning `LocatorPathElt::PatternMatch` found or not break something on test suit?

Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp]: I didn't fully understand what you just said,
I never thought it would be this easy.
Fixed it and the test case seems to pass.

Thanks for the code suggestions as well.

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp]: Awesome, if all tests pass looks like we don't need the more specific check.
We may only need to remove a couple of redundant logic in there =]

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2075]: This is redundant because we know if we found an element this is always true at this point.

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2071]: The new logic already covers the first case as well
```suggestion
      auto pathElement = llvm::find_if(path, [](LocatorPathElt &elt) {
          return elt.is<LocatorPathElt::PatternMatch>();
      });
      return pathElement != path.end();
```
So this may be enough?

Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp]: Yes👏 In my environment test seems passed.

**Comment by takeshi-1000** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
Only committer can trigger CI, so let us help you with that.

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2060]: This dropping shouldn't be needed anymore now that we only look for pattern match element.

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2062]: We should also update both comments =]

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp:2066]: There is a better llvm function for this - llvm::is_contained which returns a bool

**Comment by xedin** [comment]
@takeshi-1000 Thanks for addressing the comments! Looks like CI is green and once you deal with the last couple of nits we have mentioned this PR should be landable next week once the branch is unlocked.

File: lib/Sema/CSSimplify.cpp
Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp:2066]: sorry to reply late. tried to fix it, but it looks redundant than current code as below.

```
std::vector<ConstraintLocator::PathElementKind> kinds;
llvm::transform(path, std::back_inserter(kinds), [](LocatorPathElt &elt) {
    return elt.getKind();
});
return llvm::is_contained(kinds, ConstraintLocator::PatternMatch);

```
Could you please let me know if you have any other good code?

File: lib/Sema/CSSimplify.cpp
Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp:2061]: Thanks.
I'm thinking of changing it to:
```
Direct pattern or sub-pattern matching of tuple element.
 e.g.`case .foo((a: 42, _)) = question`
```
Do you have any suggestions? 👀

Comment (LucianoPAlmeida) [reply lib/Sema/CSSimplify.cpp:2061]: Sounds good!

File: lib/Sema/CSSimplify.cpp
Comment (LucianoPAlmeida) [inline lib/Sema/CSSimplify.cpp:2060]: We should remove this all and maybe `!path.empty()` condition is redundant since `llvm::find_if` would not find anything if the path is empty, so we can remove it as well?

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [reply lib/Sema/CSSimplify.cpp:2063]: Ah `PatternMatch` requires a `pattern` argument, so attempting to match it via llvm::is_contained wouldn't work :/ Okay, let's keep it as `find_if` then.

File: lib/Sema/CSSimplify.cpp
Comment (takeshi-1000) [reply lib/Sema/CSSimplify.cpp]: also seems fine without this :)

**Review by LucianoPAlmeida — COMMENTED** [review]
LGTM!

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please test

File: lib/Sema/CSSimplify.cpp
Comment (xedin) [inline lib/Sema/CSSimplify.cpp:2058]: This whole comment could be removed, it’s only partially correct now anyway, since the logic below handles both direct and sub-pattern matching.

**Review by xedin — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility debug

**Comment by xedin** [comment]
Thank you very much, @takeshi-1000!

**Comment by xedin** [comment]
I'll cherry-pick this into 5.8 branch tomorrow.

**Comment by takeshi-1000** [comment]
@xedin 
Thank you for confirming, also thank you very much for your help too, @LucianoPAlmeida!

Sorry after merging the PR, but can I check one thing? 
Is this degression caused by this improvement? (https://forums.swift.org/t/pitch-enable-multi-statement-closure-parameter-result-type-inference/52619)

**Pull request merged**
